text/3516-public-private-dependencies.md

```diff
@@ -0,0 +1,425 @@
+`libcore` and `libstd` in Rust itself.
+
+## cargo
+
+A new dependency field, `public = <bool>` will be added that defaults to `false`.
+This field can be specified in `workspace.dependencies` and be overridden when `workspace = true` is in a dependency.
+When building a `lib`, Cargo will use the `priv` modifier with `--extern` for all private dependencies.
+What is private is what is left after recursively walking public dependencies (`public = true`).
+For other [`crate-type`s](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#the-crate-type-field) (e.g. `bin`),
+we'll tell rustc that all dependencies are public to reduce noise from inaccessible `public` items.
+
+Cargo will not force a `rust-version` bump when using this feature as someone
+building with an old version of cargo depending on packages that set `public =
+true` will not start to fail when upgrading to new versions of cargo.
+
+`cargo add` will gain `--public <bool>` flags to control this field.
+When adding a dependency today, the version requirement is reused from other dependency tables within your manifest.
+With this RFC, that will be extended to also checking your dependencies for any `public` dependencies, and reusing their version requirement.
+This would be most easily done by having the field in the Index but `cargo add` could also read the `.crate` files as a fallback.
+
+## crates.io
+
+Crates.io should show public dependencies more prominently than private ones.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It might not be clear how to resolve the warning/error, as it's emitted
+by rustc but is resolved by changing information in the build system,
+generally, but not always, cargo.
+As a last resort, we could put a hack in cargo to intercept the lint and emit a
+new version of it that explains things in terms of cargo.
+
+There are risks with the `cargo fix` approach as it requires us to take a non-machine applicable lint,
+parsing out the information we need to identify the corresponding `Cargo.toml`,
+and translate it into a change for `Cargo.toml`.
+
+In the case where you depend on `foo = "300"`, there isn't a way to clarify that what is public is actually from `foo-core = "1"` without explicitly depending on it.
+
+This doesn't cover the case where a dependency is public only if a feature is enabled.
+
+The warning/error is emitted even when a `pub` item isn't accessible.
+We at least reduced the impact of this by not marking dependencies as private for `crate-type=["bin"]`.
+
+You can't definitively lint when a `public = true` is unused since it may depend on which platform or features.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Misc
+
+- `Cargo.toml`: instead of `public` (like [RFC 1977]), we could name the field `pub` (named after the [Rust keyword](https://doc.rust-lang.org/reference/visibility-and-privacy.html) or name the field `visibility = "<public|private>"`
```

Comment:
```suggestion
- `Cargo.toml`: instead of `public` (like [RFC 1977]), we could name the field `pub` (named after the [Rust keyword](https://doc.rust-lang.org/reference/visibility-and-privacy.html)) or name the field `visibility = "<public|private>"`
```